lib/puppet/confine/boolean.rb

```diff
@@ -0,0 +1,56 @@
+require 'puppet/confine'
+
+# Common module for the Boolean confines. It currently
+# contains just enough code to implement PUP-9336.
+class Puppet::Confine
+  module Boolean
+    # Returns the passing value for the Boolean confine. This should
+    # just be the class name.
+    def passing_value
+      return @passing_value unless @passing_value.nil?
+
+      class_name = self.class.to_s.split('::').pop.downcase.to_sym
+
+      @passing_value = class_name == :true ? true : false
+    end
+
+    # The Boolean confines 'true' and 'false' let the user specify
+    # two types of values:
+    #     * A lambda for lazy evaluation. This would be something like
+    #         confine :true => lambda { true }
+    #
+    #     * A single Boolean value, or an array of Boolean values. This would
+    #     be something like
+    #         confine :true => true OR confine :true => [true, false, false, true]
+    #
+    # This override distinguishes between the two cases.
+    def values
+      # Note that Puppet::Confine's constructor ensures that @values
+      # will always be an array, even if a lambda's passed in. This is
+      # why we have the length == 1 check.
+      unless @values.length == 1 && @values.first.respond_to?(:call)
```

Comment:
This is mostly thinking out loud here.

Would the `values` method be cleaner if, instead of saying "you can only pass in one lambda/block/proc", it iterated over everything passed in, and `#call`ed anything that supported it?

This does mean that you could mix & match with potentially multiple lambda/block/proc & static true/false values. I'm not sure if that's actually a good thing or not in practice, since you could effectively do any version of that with a single lambda/block/proc (though more complex than any of the multiple ones would be).

I'm not sure if this is an overall win, or even a good idea. Just something to think about.

Reply:
Yeah, I was thinking about that too initially and decided against it for the reason you mentioned (that it's weird mixing in lambdas with static true/false values). It would make the code a bit cleaner and we can still cache stuff if we need to, so I am happy to change the PR to support that if desired.